File: apps/launcher-api/src/lib/server/lease.ts
import { and, eq, sql } from 'drizzle-orm';
import { getDb } from './db';
import { docLeases, type DocLease } from './db/schema';

/**
 * Soft, cooperative edit lease — the Postgres half of
 * `docs/design/multi-user-concurrency.md` Phase 2. A lease is a COORDINATION
 * HINT, never authz: `toolScope.gate()` stays the real security boundary. It lets
 * two people on the same `(client, project)` avoid clobbering each other; R2
 * `If-Match` remains the correctness floor for everything a lease can't cover
 * (a stale tab whose lease expired, a takeover mid-flight, the Python tools).
 *
 * Keyed per-project for now — `(toolId, clientKey, projectKey, docKey)`, with
 * `docKey` = the tool's single project doc (pass the tool id when a tool has one
 * doc). Adjudication is done by the DATABASE via one conditional upsert, so two
 * racing acquires can never both win. Expiry is the backstop; explicit takeover
 * from the UI is the plan (a crashed tab must never permanently wedge a doc).
 */

/** How often a holder should re-heartbeat to keep the lease alive. */
export const LEASE_HEARTBEAT_MS = 10_000;
/** How long a lease survives without a heartbeat before it becomes takeable. */
export const LEASE_TTL_MS = 45_000;
const LEASE_TTL_SECONDS = LEASE_TTL_MS / 1000;

export interface LeaseKey {
	toolId: string;
	clientKey: string;
	projectKey: string;
	docKey: string;
}

export interface LeaseHolder {
	userId: string;
	sessionId: string;
}

/** Who currently holds a lease we could not take, for the read-only banner. */
export interface HeldBy {
	userId: string;
	sessionId: string;
	acquiredAt: Date;
	heartbeatAt: Date;
	expiresAt: Date;
}

export type AcquireResult =
	| { held: true; lease: DocLease }
	| { held: false; heldBy: HeldBy; activeAgoMs: number };

export type HeartbeatResult =
	| { held: true; lease: DocLease }
	| { held: false; heldBy: HeldBy | null };

/**
 * The takeability predicate, kept as a pure function so it can be unit-tested
 * directly AND so it reads identically to the SQL `setWhere` below (which is a
 * faithful transcription — an existing lease is takeable when there is none, when
 * it has expired, or when we already hold it). Keep the two in lock-step: the SQL
 * is the source of truth in production, this mirror is what the offline fixture
 * adjudicates against.
 */
export function isTakeable(
	existing: Pick<DocLease, 'holderUserId' | 'holderSessionId' | 'expiresAt'> | null,
	holder: LeaseHolder,
	nowMs: number,
): boolean {
	if (!existing) return true;
	if (existing.expiresAt.getTime() < nowMs) return true;
	return isSameHolder(existing, holder);
}

/** Whether an existing lease row is held by this exact `(userId, sessionId)`. */
export function isSameHolder(
	existing: Pick<DocLease, 'holderUserId' | 'holderSessionId'>,
	holder: LeaseHolder,
): boolean {
	return existing.holderUserId === holder.userId && existing.holderSessionId === holder.sessionId;
}

function heldBy(row: DocLease): HeldBy {
	return {
		userId: row.holderUserId,
		sessionId: row.holderSessionId,
		acquiredAt: row.acquiredAt,
		heartbeatAt: row.heartbeatAt,
		expiresAt: row.expiresAt,
	};
}

function keyMatch(key: LeaseKey) {
	return and(
		eq(docLeases.toolId, key.toolId),
		eq(docLeases.clientKey, key.clientKey),
		eq(docLeases.projectKey, key.projectKey),
		eq(docLeases.docKey, key.docKey),
	);
}

async function readLease(key: LeaseKey): Promise<DocLease | null> {
	const [row] = await getDb().select().from(docLeases).where(keyMatch(key)).limit(1);
	return row ?? null;
}

const nowSql = sql`now()`;
const expiresSql = sql`now() + ${LEASE_TTL_SECONDS} * interval '1 second'`;

/**
 * Conditional upsert — the DB adjudicates. Insert if absent; on conflict, take
 * over ONLY if the current lease is expired or already ours (`setWhere`), else
 * leave the live holder untouched and return no row. One statement, so two racing
 * acquires can't both win. `acquiredAt` resets only when the holder changes.
 */
export async function acquire(key: LeaseKey, holder: LeaseHolder): Promise<AcquireResult> {
	const sameHolderSql = sql`${docLeases.holderUserId} = excluded.holder_user_id and ${docLeases.holderSessionId} = excluded.holder_session_id`;

	const [granted] = await getDb()
		.insert(docLeases)
		.values({
			...key,
			holderUserId: holder.userId,
			holderSessionId: holder.sessionId,
			acquiredAt: nowSql,
			heartbeatAt: nowSql,
			expiresAt: expiresSql,
		})
		.onConflictDoUpdate({
			target: [docLeases.toolId, docLeases.clientKey, docLeases.projectKey, docLeases.docKey],
			set: {
				holderUserId: holder.userId,
				holderSessionId: holder.sessionId,
				acquiredAt: sql`case when ${sameHolderSql} then ${docLeases.acquiredAt} else now() end`,
				heartbeatAt: nowSql,
				expiresAt: expiresSql,
			},
			setWhere: sql`${docLeases.expiresAt} < now() or (${sameHolderSql})`,
		})
		.returning();

	if (granted) return { held: true, lease: granted };

	// The upsert's `setWhere` was false: a live, other holder owns it. Report them.
	const current = await readLease(key);
	// Vanished between the upsert and this read (released/expired) — retry once.
	if (!current) return acquire(key, holder);
	return {
		held: false,
		heldBy: heldBy(current),
		activeAgoMs: Math.max(0, Date.now() - current.heartbeatAt.getTime()),
	};
}

/**
 * Extend the lease only while WE still hold it. If it was taken over (a different
 * holder), no row matches and we return not-held so the client can flip to
 * read-only. A same-holder heartbeat re-extends even past expiry (nobody took it).
 */
export async function heartbeat(key: LeaseKey, holder: LeaseHolder): Promise<HeartbeatResult> {
	const [row] = await getDb()
		.update(docLeases)
		.set({ heartbeatAt: nowSql, expiresAt: expiresSql })
		.where(
			and(
				keyMatch(key),
				eq(docLeases.holderUserId, holder.userId),
				eq(docLeases.holderSessionId, holder.sessionId),
			),
		)
		.returning();

	if (row) return { held: true, lease: row };
	const current = await readLease(key);
	return { held: false, heldBy: current ? heldBy(current) : null };
}

/** Best-effort release; only deletes a lease we still hold. Expiry is the backstop. */
export async function release(key: LeaseKey, holder: LeaseHolder): Promise<void> {
	await getDb()
		.delete(docLeases)
		.where(
			and(
				keyMatch(key),
				eq(docLeases.holderUserId, holder.userId),
				eq(docLeases.holderSessionId, holder.sessionId),
			),
		);
}

/**
 * Force-acquire regardless of the current holder — an explicit user action that
 * is ALWAYS allowed, because a lease must never permanently wedge a doc. Recorded
 * plainly as a fresh acquisition (`acquiredAt` resets).
 */
export async function takeover(key: LeaseKey, holder: LeaseHolder): Promise<{ lease: DocLease }> {
	const [row] = await getDb()
		.insert(docLeases)
		.values({
			...key,
			holderUserId: holder.userId,
			holderSessionId: holder.sessionId,
			acquiredAt: nowSql,
			heartbeatAt: nowSql,
			expiresAt: expiresSql,
		})
		.onConflictDoUpdate({
			target: [docLeases.toolId, docLeases.clientKey, docLeases.projectKey, docLeases.docKey],
			set: {
				holderUserId: holder.userId,
				holderSessionId: holder.sessionId,
				acquiredAt: nowSql,
				heartbeatAt: nowSql,
				expiresAt: expiresSql,
			},
		})
		.returning();

	return { lease: row };
}

File: services/_shared/iw_common/lease.py
"""A soft lease over one document, for the Python tools.

The launcher's TypeScript half (`$lib/server/lease.ts`, `doc_leases`, `POST
/api/lease`) answers "which PERSON has this doc open" and is adjudicated by
Postgres. This is the same lease shape for a different collision: **two CONTAINERS
of one service**, during the overlap of a rolling Railway deploy. The names, the
timings and the semantics are copied from that module deliberately — a later move
onto `doc_leases` should be a transport swap, not a rewrite.

WHERE IT LIVES — a departure from the house rule, recorded in
`docs/design/multi-user-concurrency.md` (Phase 3 amendment). That rule says a lease
belongs in Postgres and never in the document it protects, because a lease inside
the doc is clobbered by the very race it exists to prevent. This lease is NOT in the
document: it is its own object, claimed with `If-None-Match: '*'` and renewed with
`If-Match`, so R2 adjudicates the claim atomically and the property the rule exists
to guarantee holds by a different mechanism. Postgres is unavailable here for
reasons that are structural rather than lazy — these services hold no launcher
session and no DB credentials, `doc_leases.holder_user_id` is a foreign key to a
real user and a container is not one, and routing the claim through the launcher
would make it fail open during precisely the deploy it exists to survive.

FAIL OPEN, ALWAYS. Every function answers "you may proceed" when the store cannot
be read or written: a lease that can wedge a render is worse than the race it
prevents. Be honest about what that costs, though — the conditional write on the
document is the floor for the DOC's state only. The other half of a double-submit
is a second `_submit` clearing the hand-off slots and deleting a render the first
one's job already uploaded, and no precondition covers that. So failing open
deliberately re-accepts the double-submit risk for as long as R2 itself is
unreachable, on the grounds that a tool which refuses to render at all is the worse
of the two failures.
"""
from __future__ import annotations

import json
import time
from dataclasses import dataclass

from . import storage

# Copied from `lease.ts:21-24`. The two halves cannot read each other's rows today —
# different stores entirely — so this is not about agreeing live; it is so that
# moving these rows into `doc_leases` later does not silently change when a lease
# is considered dead.
LEASE_HEARTBEAT_MS = 10_000
LEASE_TTL_MS = 45_000


@dataclass(frozen=True)
class LeaseKey:
    """The composite key `doc_leases` uses as its PRIMARY KEY, in the same order."""
    tool_id: str
    client_key: str
    project_key: str
    doc_key: str

    def path(self) -> str:
        return (f"{self.client_key}/{self.project_key}/_leases/"
                f"{self.tool_id}/{self.doc_key}.json")


@dataclass(frozen=True)
class LeaseHolder:
    """Who is asking. `session_id` is what actually distinguishes two claimants
    here — one container id per process — mirroring `holderSessionId`, which the
    schema describes as the holder's server-side session id rather than a person."""
    user_id: str
    session_id: str


def _now_ms() -> int:
    """Note the one property this does NOT copy from `lease.ts`: there, every
    timestamp comes from a single Postgres `now()`, so a TTL means the same thing to
    every claimant. Here `expiresAt` is written by one container's clock and read by
    another's, so the TTL is only ever as good as their clock agreement (NTP on
    Railway, so milliseconds — but it is an assumption, not a guarantee)."""
    return int(time.time() * 1000)


def _is_transient_conflict(e: Exception) -> bool:
    """R2's 409 `ConditionalRequestConflict`: two conditional writes raced, try
    again. Distinct from the 412 `storage.Conflict` means, which is an answer."""
    resp = getattr(e, "response", None)
    if not isinstance(resp, dict):
        return False
    return (resp.get("ResponseMetadata", {}).get("HTTPStatusCode") == 409
            or str(resp.get("Error", {}).get("Code") or "")
            == "ConditionalRequestConflict")


def is_same_holder(row: dict | None, holder: LeaseHolder) -> bool:
    """`lease.ts:74-79`. Both halves, because one user in two containers is exactly
    the case this exists to separate."""
    if not row:
        return False
    return (str(row.get("holderUserId") or "") == holder.user_id
            and str(row.get("holderSessionId") or "") == holder.session_id)


def is_takeable(row: dict | None, holder: LeaseHolder, now_ms: int) -> bool:
    """`lease.ts:63-71`, kept pure for the same reason it is pure there: it has to
    read identically to the condition the store enforces, and that is only checkable
    if it can be fixtured on its own.

    Takeable when nothing holds it, when what holds it has expired, or when it is
    already ours.
    """
    if not row:
        return True
    if is_same_holder(row, holder):
        return True
    # `<`, not `<=`, because the SQL twin is `expires_at < now()` — these
    # predicates exist to read identically to the condition the store enforces.
    return int(row.get("expiresAt") or 0) < now_ms


def _row(key: LeaseKey, holder: LeaseHolder, acquired_ms: int, now_ms: int,
         ttl_ms: int) -> dict:
    return {
        "toolId": key.tool_id,
        "clientKey": key.client_key,
        "projectKey": key.project_key,
        "docKey": key.doc_key,
        "holderUserId": holder.user_id,
        "holderSessionId": holder.session_id,
        "acquiredAt": acquired_ms,
        "heartbeatAt": now_ms,
        "expiresAt": now_ms + ttl_ms,
    }


def read(key: LeaseKey) -> tuple[dict | None, str | None]:
    """The stored row and its ETag, or `(None, None)` when there genuinely is none.

    RAISES `storage.ObjectUnreadable` when the store could not be asked. Collapsing
    that into "nothing is there" is the mistake `get_with_etag` exists to prevent:
    a caller would take the create-if-absent branch over a live row, be refused by
    the precondition, and report "somebody else holds this" — failing CLOSED in a
    module whose whole contract is to fail open. Callers decide what unreadable
    means for them; here it means "proceed".
    """
    got = storage.get_with_etag(key.path())
    if not got:
        return None, None
    body, etag = got
    try:
        row = json.loads(body)
    except ValueError:
        return None, etag  # a corrupt row is a takeable one
    return (row if isinstance(row, dict) else None), etag


def acquire(key: LeaseKey, holder: LeaseHolder, *,
            ttl_ms: int = LEASE_TTL_MS) -> bool:
    """Take the lease if it is free, expired, or already ours. True when we hold it.

    `lease.ts:114` does this as ONE conditional upsert so the DB adjudicates and two
    racing acquires can never both win. Here R2 adjudicates instead: the write
    carries `If-None-Match: '*'` when we believe nothing holds it and `If-Match`
    when we believe we are replacing what we just read, so a claim that raced
    another container loses on the precondition rather than by overwriting it.
    """
    try:
        row, etag = read(key)
    except Exception:  # noqa: BLE001 — a store we cannot ask holds nobody
        return True
    now = _now_ms()
    if not is_takeable(row, holder, now):
        return False
    # `acquiredAt` survives a renewal by the same holder and resets on a change of
    # hands — `lease.ts:132`, where it is what "held since" is read from.
    acquired = int(row.get("acquiredAt") or now) if is_same_holder(row, holder) \
        else now
    body = json.dumps(_row(key, holder, acquired, now, ttl_ms),
                      indent=2).encode("utf-8")
    for attempt in range(3):
        try:
            if row is None and etag is None:
                storage.put(key.path(), body, "application/json",
                            if_none_match="*")
            else:
                storage.put(key.path(), body, "application/json",
                            if_match=etag or "*")
            return True
        except storage.Conflict:
            return False      # somebody claimed it between our read and our write
        except Exception as e:  # noqa: BLE001
            # R2 answers 409 `ConditionalRequestConflict` when two conditional
            # writes to one key are in flight at the same instant — which is
            # precisely two containers claiming one session. It is transient and
            # must be RETRIED: treating it as success would let both claimants
            # believe they won, and the destructive half of a double-submit (a
            # second `_submit` deleting the first job's uploaded render) has no
            # precondition protecting it.
            if _is_transient_conflict(e) and attempt < 2:
                time.sleep(0.15 * (attempt + 1))
                continue
            if _is_transient_conflict(e):
                return False  # still contended after retries: we did not get it
            return True       # a store we cannot write at all holds nobody
    return False


def heartbeat(key: LeaseKey, holder: LeaseHolder, *,
              ttl_ms: int = LEASE_TTL_MS) -> bool:
    """Extend a lease we still hold. False once somebody has taken it from us —
    `lease.ts:158`, where that is the signal the client flips read-only on.

    A same-holder heartbeat re-extends even past expiry: the alternative is a holder
    that is plainly alive being told to stop by a clock. A foreign row that has
    EXPIRED is not a takeover either — it is a lease nobody holds — so a dispatcher
    whose own lease lapsed under it can take it back rather than standing down for
    good."""
    try:
        row, _etag = read(key)
    except Exception:  # noqa: BLE001 — unreadable holds nobody
        return True
    if not is_takeable(row, holder, _now_ms()):
        return False
    return acquire(key, holder, ttl_ms=ttl_ms)


def release(key: LeaseKey, holder: LeaseHolder) -> None:
    """Give it up. Best-effort — expiry is the backstop (`lease.ts:177`), so a
    delete that fails costs one TTL, never correctness.

    Reads first and skips a row somebody else holds, so a release arriving after a
    takeover does not delete the new holder's lease. That check and the delete are
    two calls, not one conditional delete, so a takeover landing between them is
    still possible; the window is milliseconds and the cost is one TTL of nobody
    holding a session that is not being rendered anyway."""
    try:
        row, _etag = read(key)
    except Exception:  # noqa: BLE001 — if we cannot read it, let it expire
        return
    if row is None or is_same_holder(row, holder):
        try:
            storage.delete(key.path())
        except Exception:  # noqa: BLE001
            pass


def takeover(key: LeaseKey, holder: LeaseHolder, *,
             ttl_ms: int = LEASE_TTL_MS) -> bool:
    """Take it regardless of who holds it. Unconditional, because a lease must never
    permanently wedge a doc (`lease.ts:194`) — here that is Stop, which has to reach
    a session whichever container owns it."""
    now = _now_ms()
    body = json.dumps(_row(key, holder, now, now, ttl_ms), indent=2).encode("utf-8")
    try:
        storage.put(key.path(), body, "application/json")
        return True
    except Exception:  # noqa: BLE001 — fail open: the stop proceeds either way
        return True


def held_by(key: LeaseKey) -> dict | None:
    """The live holder's row, or None when nothing holds it — the read a caller
    makes to decide whether to defer, never to decide whether to be allowed."""
    row, _etag = read(key)          # raises when the store cannot be asked
    if row and int(row.get("expiresAt") or 0) > _now_ms():
        return row
    return None
